src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,506 @@
+        self.projection_dim = config.projection_dim
+        if self.projection_dim > 0:
+            self.encode_proj = nn.Linear(self.bert_model.config.hidden_size, config.projection_dim)
+        self.init_weights()
+
+    def forward(
+        self, input_ids: Tensor, token_type_ids: Optional[Tensor], attention_mask: Optional[Tensor]
+    ) -> Tuple[Tensor, ...]:
+        outputs = self.bert_model(
+            input_ids=input_ids,
+            token_type_ids=token_type_ids,
+            attention_mask=attention_mask,
+            output_hidden_states=True,
+        )
+        sequence_output, pooled_output, hidden_states = outputs
+        pooled_output = sequence_output[:, 0, :]
+        if self.projection_dim > 0:
+            pooled_output = self.encode_proj(pooled_output)
+        return sequence_output, pooled_output, hidden_states
+
+    @property
+    def embeddings_size(self) -> int:
+        if self.projection_dim > 0:
+            return self.encode_proj.out_features
+        return self.bert_model.config.hidden_size
+
+    def init_weights(self):
+        self.bert_model.init_weights()
+
+
+class DPRSpanPredictor(PreTrainedModel):
+    def __init__(self, config: DPRConfig):
+        super().__init__(config)
+        self.encoder = DPREncoder(config)
+        self.qa_outputs = nn.Linear(self.encoder.embeddings_size, 2)
+        self.qa_classifier = nn.Linear(self.encoder.embeddings_size, 1)
+        self.init_weights()
+
+    def forward(self, input_ids: Tensor, attention_mask: Tensor):
+        # notations: N - number of questions in a batch, M - number of passages per questions, L - sequence length
+        N, M, L = input_ids.size()
+        # resize as matrices
+        input_ids = input_ids.view(N * M, L)
+        attention_mask = attention_mask.view(N * M, L)
+        # feed encoder
+        sequence_output, *_ = self.encoder(input_ids, None, attention_mask)
+        # compute logits
+        logits = self.qa_outputs(sequence_output)
+        start_logits, end_logits = logits.split(1, dim=-1)
+        start_logits = start_logits.squeeze(-1)
+        end_logits = end_logits.squeeze(-1)
+        relevance_logits = self.qa_classifier(sequence_output[:, 0, :])
+        # resize and return
+        return start_logits.view(N, M, L), end_logits.view(N, M, L), relevance_logits.view(N, M)
+
+    def init_weights(self):
+        self.encoder.init_weights()
+
+
+##################
+# PreTrainedModel
+##################
+
+
+CheckpointState = collections.namedtuple(
+    "CheckpointState", ["model_dict", "optimizer_dict", "scheduler_dict", "offset", "epoch", "encoder_params"]
+)
+
+
+def load_states_from_checkpoint(model_file: str) -> CheckpointState:
+    logger.info("Reading saved model from %s", model_file)
+    state_dict = torch.load(model_file, map_location=lambda s, l: default_restore_location(s, "cpu"))
+    logger.info("model_state_dict keys %s", state_dict.keys())
+    return CheckpointState(**state_dict)
```

Comment:
Can be removed from here since they were moved to the standalone conversion script